dns.go

```diff
@@ -0,0 +1,113 @@
+package gubernator
+
+import (
+	"context"
+	"time"
+
+	"github.com/mailgun/holster/v4/setter"
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type DNSPoolConfig struct {
+	// (Required) The FQDN that should resolve to gubernator instance ip addresses
+	FQDN string
+
+	// (Required) Own GRPC address
+	OwnAddress string
+
+	// (Required) Called when the list of gubernators in the pool updates
+	OnUpdate UpdateFunc
+
+	Logger logrus.FieldLogger
+}
+
+type DNSPool struct {
+	log    logrus.FieldLogger
+	conf   DNSPoolConfig
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewDNSPool(conf DNSPoolConfig) (*DNSPool, error) {
+	setter.SetDefault(&conf.Logger, logrus.WithField("category", "gubernator"))
+
+	if conf.OwnAddress == "" {
+		return nil, errors.New("Advertise.GRPCAddress is required")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	pool := &DNSPool{
+		log:    conf.Logger,
+		conf:   conf,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	go pool.task()
+	return pool, nil
+}
+
+func peer(ip string, self string, ipv6 bool) PeerInfo {
+
+	if ipv6 {
+		ip = "[" + ip + "]"
+	}
+	grpc := ip + ":81"
+	return PeerInfo{
+		DataCenter:  "",
+		HTTPAddress: ip + ":80",
+		GRPCAddress: grpc,
+		IsOwner:     grpc == self,
+	}
+
+}
+
+func (x *DNSPool) task() {
+	for {
+		var delay uint32 = 10
+		config, _ := dns.ClientConfigFromFile("/etc/resolv.conf")
+		if len(config.Servers) > 0 {
+			c := new(dns.Client)
+			c.SingleInflight = true
+			var update []PeerInfo
+			m4 := new(dns.Msg)
+			m4.SetQuestion(dns.Fqdn(x.conf.FQDN), dns.TypeA)
+			r4, _, err4 := c.Exchange(m4, config.Servers[0]+":"+config.Port)
+			m6 := new(dns.Msg)
+			m6.SetQuestion(dns.Fqdn(x.conf.FQDN), dns.TypeAAAA)
+			r6, _, err6 := c.Exchange(m6, config.Servers[0]+":"+config.Port)
```

Comment:
What if you queried both of these individually in separate functions then merged the result. Then you could pick the longest delay from the TTL (or shortest) and use that. (Might optionally provide an configurable override) 

psudocode:
```
aPeers, aTTL, err := lookupARecords()
if err != nil {
}
aaaPeers, aaaTTL, err := lookupAAARecords()
if err != nil {
}
// merge aPeers and aaaPeers
peers = append(peers, aPeers, aaaPeers)
if aTTL > aaaTTL {
   delay = aTTL
}
```

Reply:
Good point, will do!

I was in a weird mindset that the FQDN would resolve to either a set of ipv4 address or a set of ipv6 addresses, but probably not both, because in a dual-stack environment, each peer may have both v4 and v6 address, and I'm not sure how `gubernator` would handle that.

Reply:
TBH, I'd rather find a higher-level DNS library, but so far that escaped my go-fu.

Reply:
maybe I'll just copy-paste https://github.com/TimothyYe/godns/blob/master/resolver/resolver.go if the licenses match up.